hera_cal/abscal.py

```diff
@@ -1291,6 +1295,280 @@ def global_phase_slope_logcal(model, data, antpos, reds=None, solver='linfit', w
+                grid[grid_indices[nn]] = weights_t_f[nn] * nmdr_t_f[nn]
+                
+            # Find grid point maximum in Fourier space corresponding to phase slope here
+            grid_ft = np.fft.fftshift(fft.fftn(grid, workers=-1)) / Ngroups            
+            max_idx = np.unravel_index(np.argmax(np.real(grid_ft)), grid_ft.shape)
+            Lambda_init = np.array([ft_freqs[kk][ii] for kk, ii in enumerate(max_idx)])
+            
+            # Refine solution with Newton's method
+            Lambda_t_f, _ = _newton_solve_real(Lambda_init, transformed_b_vecs, nmdr_t_f, weights_t_f, 1e-8)
+            
+            # Store results
+            Lambda_sol[i_t, i_f] = Lambda_t_f
+            Z_sol[i_t, i_f] = _eval_Z(Lambda_t_f, transformed_b_vecs, nmdr_t_f, weights_t_f)
+    
+    return Lambda_sol, Z_sol
+
+def _put_transformed_array_on_integer_grid(transformed_antpos, tol=1e-8, max_numerator=100):
+    """
+    Multiplies an antenna position dictionary by a rational number to make all positions integers, up to tol.
+    Searches multipliers up to max_numerator, then looks for the largest denominator to compress the array.
+    Function modifies the input dictionary in place.
+
+    Parameters:
+    -----------
+    transformed_antpos: dict
+        Dictionary of antenna positions in the transformed coordinate system.
+    tol: float
+        Tolerance for the positions to be considered integers.
+    max_numerator: int
+        Maximum numerator to search for a multiplier.
+    """
+    # Loop through dimensions of the transformed antpos
+    for dim in range(len(list(transformed_antpos.values())[0])):
+        this_dim_positions = np.array([ap[dim] for ap in transformed_antpos.values()])
+        
+        # If any position not on integer grid
+        if np.max(np.abs(np.rint(this_dim_positions) - this_dim_positions)) > tol:
+            
+            # Look for multipliers that put every antenna on an integer grid
+            for numerator in range(2, max_numerator + 1):
+                if np.max(np.abs(np.rint(this_dim_positions * numerator) - this_dim_positions * numerator)) < tol:
+                    break
+                assert numerator < max_numerator, 'Could not find a reasonable multiplier to put this array on an integer grid.'
+            
+            # Look for divisors compress the array but keep every antenna on an integer grid
+            for denominator in range(max_numerator, 0, -1):
+                if np.max(np.abs(np.rint(this_dim_positions * numerator / denominator) - this_dim_positions * numerator / denominator)) < tol:
+                    break
+            
+            # Apply ratio found
+            for ap in transformed_antpos:
+                transformed_antpos[ap][dim] *= (numerator / denominator)
+
+def complex_phase_abscal(data, model, reds, data_bls, model_bls):
+    """
+    Calculates gains that would absolute calibrate already redundantly-calibrated data.  
```

Comment:
This should clarify that it only performs absolute calibration of phase slopes, not amplitudes.

Reply:
I'll make that change

Reply:
Done.